astropy/table/index.py

```diff
@@ -897,17 +919,15 @@ def __setitem__(self, key, value):
         value : New values of the row elements.
                 Can be a list of tuples/lists to update the row.
         """
-        rows = self._get_rows(key)
-        if len(rows) == 0:  # no matches found
-            raise KeyError(f"No matches found for key {key}")
-        elif len(rows) == 1:  # single row
-            self.table[rows[0]] = value
-        else:  # multiple rows
+        rows = self._get_row_idxs_as_list_or_int(key)
```

Comment:
Note, the `No matches found for key` exception is already raised upstream in `_get_rows`. The conditional handling in the original code is now in `_get_row_idxs_as_list_or_int`.